### PR TITLE
Add `CustomTableHeaderLabel` formatting to differentiate table header labels that are not property names

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Add `$PSStyle.Formatting.CustomTableHeaderLabel` to enable differentiation in `Format-Table` output where a table header label isn't an actual property name.  Usually these are either abbreviated, calculated, or has units added.  Users can get confused because they expect it to be a property name and can't use dot-notation to access it.  The default format is the same color as the table header but use italics.

Since decoration is absolute and not relative (additive), to make this work correctly for each table header which IS a property member we need to wrap it with the normal decoration.  This results in extra characters being emitted.

## PR Context

Brought up in https://twitter.com/JustinWGrote/status/1524882725781377024

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): `PSCustomTableHeaderLabelDecoration `
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: [8815](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8815)
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs [hunk 1]
```diff
@@ -346,6 +346,17 @@ public string TableHeader
 
             private string _tableHeader = "\x1b[32;1m";
 
+            /// <summary>
+            /// Gets or sets the style for modified table headers.
```
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs:350]: Why do we name this as "headers" if really they are "header labels"?

**Comment by SteveL-MSFT** [comment]
Need to rework this as it doesn't work when a header gets split across multiple lines due to width restrictions

File: src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs:350]: I don't think the distinction is important for end users

Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/PSStyle.cs]: On second thought, maybe it makes sense as `CustomTableHeaderLabel`

**Review by iSazonov — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
Made a change based on @JamesWTruher feedback where the style includes a background color.  So the change keeps the whitespace separation between columns.

image

**Review by JamesWTruher — APPROVED** [review]
i think this is fine

File: src/System.Management.Automation/FormatAndOutput/common/FormattingObjects.cs [hunk 2]
```diff
@@ -207,6 +207,7 @@ internal sealed partial class TableColumnInfo : FormatInfoData
         public int alignment = TextAlignment.Left;
         public string label = null;
         public string propertyName = null;
+        internal bool HeaderMatchesProperty = true;
```
Comment (dkaszews) [inline src/System.Management.Automation/FormatAndOutput/common/FormattingObjects.cs:210]: Why not just `public`?

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs [hunk 3]
```diff
@@ -390,17 +383,51 @@ private string[] GenerateTableRow(string[] values, ReadOnlySpan<int> alignment,
...
                     if (col == lastColWithContent[row] && screenRows > 1)
                     {
-                        sb.Append(scArray[col][row].TrimEnd());
+                        value = scArray[col][row].TrimEnd();
                     }
                     else
                     {
-                        sb.Append(scArray[col][row]);
+                        value = scArray[col][row];
+                    }
```
Comment (dkaszews) [inline src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs:400]: Hoist assignment outside condition:

```csharp
string value = scArray[col][row];

// if the column is the last column with content, we need to trim trailing whitespace, unless there is only one row
if (col == lastColWithContent[row] && screenRows > 1)
{
    value = value.TrimEnd();
}
```

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs [hunk 4]
```diff
@@ -390,17 +383,51 @@ private string[] GenerateTableRow(string[] values, ReadOnlySpan<int> alignment,
...
+                    if (isHeader)
+                    {
+                        if (_si.columnInfo[col].HeaderMatchesProperty || !ExperimentalFeature.IsEnabled(ExperimentalFeature.PSCustomTableHeaderLabelDecoration))
+                        {
+                            sb.Append(PSStyle.Instance.Formatting.TableHeader);
+                        }
+                        else
+                        {
+                            // columns always start with a leading whitespace, so we insert the decoration right after that
+                            if (value.Length > 0)
```
Comment (dkaszews) [inline src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs:411]: `else if`

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs [hunk 5]
```diff
@@ -390,17 +383,51 @@ private string[] GenerateTableRow(string[] values, ReadOnlySpan<int> alignment,
...
+                            {
+                                if (col == 0)
+                                {
+                                    value = value.Insert(0, PSStyle.Instance.Formatting.CustomTableHeaderLabel);
+                                }
+                                else
+                                {
+                                    // after the first column, each additional column starts with a whitespace for separation
+                                    value = value.Insert(1, PSStyle.Instance.Formatting.CustomTableHeaderLabel);
+                                }
```
Comment (dkaszews) [inline src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs:421]:
```suggestion
                                // after the first column, each additional column starts with a whitespace for separation
                                value = value.Insert(col == 0 ? 0 : 1, PSStyle.Instance.Formatting.CustomTableHeaderLabel);
```

**Review by dkaszews — CHANGES_REQUESTED** [review]
Please provide example screenshots before and after change

File: src/System.Management.Automation/FormatAndOutput/common/FormattingObjects.cs (hunk 2 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/FormattingObjects.cs:210]: I guess if we want to follow the other members, there's no reason not to make it public

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 3 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs:400]: good suggestion

**Review by dkaszews — APPROVED** [review]

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs [hunk 6]
```diff
@@ -390,17 +383,36 @@ private string[] GenerateTableRow(string[] values, ReadOnlySpan<int> alignment,
...
                     {
-                        sb.Append(scArray[col][row].TrimEnd());
+                        value = value.TrimEnd();
                     }
-                    else
+
+                    if (isHeader)
                     {
-                        sb.Append(scArray[col][row]);
+                        if (_si.columnInfo[col].HeaderMatchesProperty || !ExperimentalFeature.IsEnabled(ExperimentalFeature.PSCustomTableHeaderLabelDecoration))
```
Comment (daxian-dbw) [inline src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs:400]:
```
|| !ExperimentalFeature.IsEnabled(ExperimentalFeature.PSCustomTableHeaderLabelDecoration))
```

Do you still need this part? The calculation of `ci.HeaderMatchesProperty` already take into account whether or not the experimental feature is enabled.

**Review by daxian-dbw — COMMENTED** [review]
Left a comment to block merging the PR for now.
~Will update once finish reviewing the rest.~ [Update] No more comments.

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 6 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs:400]: You are correct, removed.

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by StartAutomating** [comment]
@SteveL-MSFT Thanks for doing this.

It seems like a _mostly_ good idea, 

Two high level thoughts:

1. Inverting the format is too visually "loud".  I'd prefer something like Yellow+Bold.  If you want to avoid "worrying" users about this, maybe "Magenta+Bold"

2. At least one "exception to the rule" should exist.

There are many times where I use a "calculated" property simply to improve the display of that property. Two good examples are:

* Colorizing a property
* Spreading content over multiple lines.

If the calculated header name does not exist in the object, that is when I'd highlight it as a calculated property.

If getting back to the core object context is too tricky at this point of execution, one could also look for the property header within the calculation script.

**Pull request merged**